eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs

```diff
@@ -0,0 +1,168 @@
+spec = describe "UTXOW" $ do
+  it "MissingRedeemers" $ do
+    let lang = eraMaxLanguage @era
+    let scriptHash = withSLanguage lang (hashPlutusScript . guessTheNumber3)
+    txIn <- produceScript scriptHash
+    let txToSubmit = mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+    let fixup tx =
+          fixupResetAddrWits $ tx & witsTxL . rdmrsTxWitsL .~ Redeemers mempty
+    let missingRedeemer = mkSpendingPurpose $ AsItem txIn
+    withPostFixup fixup $
+      submitFailingTx
+        txToSubmit
+        [ injectFailure $
+            MissingRedeemers [(missingRedeemer, scriptHash)]
+        , injectFailure $
+            CollectErrors [NoRedeemer missingRedeemer]
+        ]
+
+  it "MissingRequiredDatums" $ do
+    let lang = eraMaxLanguage @era
+    let scriptHash = withSLanguage lang (hashPlutusScript . guessTheNumber3)
+    txIn <- produceScript scriptHash
+    let txToSubmit = mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+    let fixup tx =
+          fixupResetAddrWits $ tx & witsTxL . datsTxWitsL .~ mempty
+    let missingDatum = hashData @era (Data (P.I 3))
+    withPostFixup fixup $
+      submitFailingTx
+        txToSubmit
+        [ injectFailure $
+            MissingRequiredDatums [missingDatum] []
+        ]
+
+  it "NotAllowedSupplementalDatums" $ do
+    let lang = eraMaxLanguage @era
+    let scriptHash = withSLanguage lang (hashPlutusScript . guessTheNumber3)
+    txIn <- produceScript scriptHash
+    let extraDatumHash = hashData @era (Data (P.I 30))
+    let extraDatum = Data (P.I 30)
+    let tx =
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . inputsTxBodyL .~ [txIn]
+            & witsTxL . datsTxWitsL .~ TxDats (Map.singleton extraDatumHash extraDatum)
+    submitFailingTx
+      tx
+      [ injectFailure $
+          NotAllowedSupplementalDatums [extraDatumHash] []
+      ]
+
+  it "PPViewHashesDontMatch" $ do
+    let lang = eraMaxLanguage @era
+    let scriptHash = withSLanguage lang (hashPlutusScript . guessTheNumber3)
+    txIn <- produceScript scriptHash
+    tx <- fixupTx $ mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+    void $ impAnn "Mismatched " $ do
+      wrongIntegrityHash <- arbitrary
+      wrongIntegrityHashTx <-
+        resetAddrWits $ tx & bodyTxL . scriptIntegrityHashTxBodyL .~ SJust wrongIntegrityHash
+      withNoFixup $
+        submitFailingTx
+          wrongIntegrityHashTx
+          [ injectFailure $
+              PPViewHashesDontMatch (SJust wrongIntegrityHash) (tx ^. bodyTxL . scriptIntegrityHashTxBodyL)
+          ]
+    void $ impAnn "Missing" $ do
+      missingIntegrityHashTx <-
+        resetAddrWits $ tx & bodyTxL . scriptIntegrityHashTxBodyL .~ SNothing
+      withNoFixup $
+        submitFailingTx
+          missingIntegrityHashTx
+          [ injectFailure $
+              PPViewHashesDontMatch SNothing (tx ^. bodyTxL . scriptIntegrityHashTxBodyL)
+          ]
+
+  it "UnspendableUTxONoDatumHash" $ do
```

Comment:
Let's get rid of:
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L324-L331
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L805-L833